File: JavaExamples/src/main/java/org/firstinspires/ftc/teamcode/examplerobot/Auto.java
package org.firstinspires.ftc.teamcode.examplerobot;

import static com.pedropathing.api.Paths.*;

import static org.firstinspires.ftc.teamcode.examplerobot.Teleop.teleop;

import static dev.frozenmilk.dairy.mercurial.continuations.Continuations.*;

import com.pedropathing.api.PoseFactory;
import com.pedropathing.math.Pose;
import com.pedropathing.paths.Path;

import org.firstinspires.ftc.teamcode.examplerobot.pedro.Constants;

import dev.frozenmilk.dairy.mercurial.ftc.MercurialFTC;
import dev.frozenmilk.dairy.mercurial.ftc.MercurialFTC.RegisterableProgram;
import dev.frozenmilk.dairy.mercurial.pedro.MercurialFollower;
import dev.frozenmilk.dairy.mercurial.processes.Fiber;
import dev.frozenmilk.dairy.mercurial.processes.Spawnable;

@SuppressWarnings("unused")
public class Auto {
    private static final PoseFactory p = PoseFactory.degrees();
    private static final Pose start = p.of(24.0, 24.0, 0.0);
    private static final Pose end = p.of(48.0, 48.0, 90.0);
    private static final Path path1 = line(start, end);
    private static final Path path2 = line(end, start);

    public static final RegisterableProgram auto = MercurialFTC.autonomous(ctx -> {
        MercurialFollower pedro = Constants.create(ctx.hardwareMap());

        pedro.localizer().setPose(start);

        Drivetrain drivetrain = new Drivetrain(pedro);
        Shooter shooter = new Shooter(ctx.hardwareMap(), pedro.localizationEventManager());

        Spawnable<?> program = expression((scope, _self) -> {
            // wait for the opmode to start
            scope.run(waitUntil(ctx::isActive));
            scope.runExec(() -> shooter.target(end));
            scope.run(waitFor.bind(seconds(10)));
            scope.runExec(shooter::idle);
            scope.run(drivetrain.followSync(path2, true));
            return noop;
        }).spawnable();

        // spawn and link
        Fiber<?> fiber = program.spawnLink();

        // yield remainder of auto to the scheduler
        ctx.dropToScheduler();

        // return into teleop
        // this will auto transition you into teleop
        // with the correctly set pose
        return teleop(pedro.localizer().pose());
    });
}

File: KotlinExamples/src/main/kotlin/org/firstinspires/ftc/teamcode/examplerobot/Auto.kt
package org.firstinspires.ftc.teamcode.examplerobot

import com.pedropathing.api.Paths.line
import com.pedropathing.api.PoseFactory
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.exec
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.expression
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.noop
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.seconds
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.waitFor
import dev.frozenmilk.dairy.mercurial.continuations.Continuations.waitUntil
import dev.frozenmilk.dairy.mercurial.ftc.MercurialFTC
import org.firstinspires.ftc.teamcode.examplerobot.pedro.Constants


private val p = PoseFactory.degrees()
private val start = p.of(24.0, 24.0, 0.0)
private val end = p.of(48.0, 48.0, 90.0)
private val path1 = line(start, end)
private val path2 = line(end, start)

@Suppress("unused")
val auto = MercurialFTC.autonomous {
    val pedro = Constants.create(hardwareMap)

    pedro.localizer.setPose(start)

    val drivetrain = Drivetrain(pedro)
    val shooter = Shooter(hardwareMap, pedro.localizationEventManager)

    val program = expression {
        // wait for the opmode to start
        !waitUntil(::isActive)
        !drivetrain.followSync(path1, true)
        !exec { shooter.target(end) }
        !(waitFor * seconds { 10.0 })
        !exec { shooter.idle() }
        !drivetrain.followSync(path2, true)
        noop
    }.spawnable()

    // spawn and link
    val fiber = program.spawnLink()

    // yield remainder of auto to the scheduler
    dropToScheduler()

    // return into teleop
    // this will auto transition you into teleop
    // with the correctly set pose
    teleop(pedro.localizer.pose())
}